src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs

```diff
@@ -509,6 +509,28 @@ private bool CompleteHandshake(ref ProtocolToken? alertToken, out SslPolicyError
                 return true;
             }
 
+            if (_selectedClientCertificate != null && !CertificateValidationPal.IsLocalCertificateUsed(_securityContext!))
```

Comment:
@wfurt I had to move this piece of code from `VerifyRemoteCertificate` up here to make the mutual authentication check work. I'm still waiting for tests to finish to confirm that I didn't break the logic on some other platform.

Reply:
On Android the problem is that we call the `VerifyRemoteCertificate` method _before_ the client certificate is sent to the server and so we cleared it before we used it.

As far as I can tell there is no side effect in the `GetRemoteCertificate` on any platform that would affect `IsLocalCertificateUsed` (which only has non-trivial implementations on Windows and Android).

I don't really understand why @wfurt chose `VerifyRemoteCertificate` to check if the client certificate was used or not. It seems mostly unrelated to what the method does and I can't see it being used anywhere. As far as I can tell, moving the code shouldn't have any impact, but I might have missed something.

Reply:
Conceptually, `VerifyRemoteCertificate` and `IsLocalCertificateUsed` are fully independent. We generally need to do this in scope of `CompleteHandshake` and it was convenient to do local & remote certificate queries at the same place. It seems like the problem here is that on Android the handshake is not completed - silent assumption other platforms have.

On Linux there is still OpenSSL validation callback. But we tell it to trust everything and we do real validation later. That _may_ change in distant future but it is working like that now. The downside is that peer may get TLS Alter or IOError _after_ it sees handshake completed.   

If all that make sense, I feel it is OK change. It is probably better to have this explicitly in CompleteHandshake to avoid confusing.

Reply:
Yes, we're now calling `VerifyRemoteCertificate` before the handshake is completed on Android and that's why I needed to make this code change. Thanks for the clarification.